src/Cli/dotnet/commands/dotnet-sln/add/Program.cs

```diff
@@ -43,27 +49,39 @@ internal class AddProjectToSolutionCommand : CommandBase
             {
                 _relativeRootSolutionFolders = null;
             }
+
+            var slnFile = _arguments.FirstOrDefault(path => path.EndsWith(".sln"));
+            if (slnFile != null)
+            {
+                string args;
+                if (_inRoot) args = $"--{SlnAddParser.InRootOption.Name} ";
+                else if (hasRelativeRoot) args = $"--{SlnAddParser.SolutionFolderOption.Name} {string.Join(" ", relativeRoot)} ";
+                else args = "";
+
+                var projectArgs = string.Join(" ", _arguments.Where(path => !path.EndsWith(".sln")));
+
+                throw new GracefulException(new string[]
+                {
+                    string.Format(CommonLocalizableStrings.SolutionArgumentMisplaced, slnFile),
+                    CommonLocalizableStrings.DidYouMean,
+                    $"  dotnet sln {slnFile} add {args}{projectArgs}"
+                });
+            }
         }
 
         public override int Execute()
         {
             SlnFile slnFile = SlnFileFactory.CreateFromFileOrDirectory(_fileOrDirectory);
 
-            var arguments = (_parseResult.ValueForArgument<IEnumerable<string>>(SlnAddParser.ProjectPathArgument) ?? Array.Empty<string>()).ToList().AsReadOnly();
-            if (arguments.Count == 0)
-            {
-                throw new GracefulException(CommonLocalizableStrings.SpecifyAtLeastOneProjectToAdd);
-            }
-
-            PathUtility.EnsureAllPathsExist(arguments, CommonLocalizableStrings.CouldNotFindProjectOrDirectory, true);
+            PathUtility.EnsureAllPathsExist(_arguments, CommonLocalizableStrings.CouldNotFindProjectOrDirectory, true);
 
-            var fullProjectPaths = arguments.Select(p =>
+            var fullProjectPaths = _arguments.Select(p =>
             {
                 var fullPath = Path.GetFullPath(p);
                 return Directory.Exists(fullPath) ?
                     MsbuildProject.GetProjectFileFromDirectory(fullPath).FullName :
                     fullPath;
-            }).ToList();
+            }).ToArray();
```

Comment:
Why is this change necessary? I think it should be reverted.

Reply:
It was part of changes I was working on trying to deal with https://github.com/dotnet/sdk/issues/18607 but it became a deeper problem than I wanted to address in this PR which is why I made a new issue.

I'll revert it and the same change in `dotnet sln remove` since its not needed for this PR.